File: FBSimulatorControl/Commands/FBSimulatorLifecycleCommands.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBSimulatorLifecycleCommands.h"

#import <CoreSimulator/SimDevice.h>

#import <AppKit/AppKit.h>

#import "FBCoreSimulatorNotifier.h"
#import "FBSimulator.h"
#import "FBSimulatorBootConfiguration.h"
#import "FBSimulatorBootStrategy.h"
#import "FBSimulatorBridge.h"
#import "FBSimulatorConfiguration+CoreSimulator.h"
#import "FBSimulatorConfiguration.h"
#import "FBSimulatorControl.h"
#import "FBSimulatorControlConfiguration.h"
#import "FBSimulatorError.h"

const int OPEN_URL_RETRIES = 2;

@interface FBSimulatorLifecycleCommands ()

@property (nonatomic, weak, readonly) FBSimulator *simulator;
@property (nonatomic, strong, readwrite, nullable) FBSimulatorHID *hid;
@property (nonatomic, strong, readwrite, nullable) FBSimulatorBridge *bridge;

@end

@implementation FBSimulatorLifecycleCommands

#pragma mark Initializers

+ (instancetype)commandsWithTarget:(FBSimulator *)target
{
  return [[self alloc] initWithSimulator:target];
}

- (instancetype)initWithSimulator:(FBSimulator *)simulator
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _simulator = simulator;

  return self;
}

#pragma mark Boot/Shutdown

- (FBFuture<NSNull *> *)boot:(FBSimulatorBootConfiguration *)configuration
{
  return [FBSimulatorBootStrategy boot:self.simulator withConfiguration:configuration];
}

#pragma mark FBPowerCommands

- (FBFuture<NSNull *> *)shutdown
{
  return [[self.simulator.set shutdown:self.simulator] mapReplace:NSNull.null];
}

- (FBFuture<NSNull *> *)reboot
{
  return [[self
    shutdown]
    onQueue:self.simulator.workQueue fmap:^(id _) {
      return [self boot:FBSimulatorBootConfiguration.defaultConfiguration];
    }];
}

#pragma mark Erase

- (FBFuture<NSNull *> *)erase
{
  return [[self.simulator.set erase:self.simulator] mapReplace:NSNull.null];
}

#pragma mark States

- (FBFuture<NSNull *> *)resolveState:(FBiOSTargetState)state
{
  return FBiOSTargetResolveState(self.simulator, state);
}

- (FBFuture<NSNull *> *)resolveLeavesState:(FBiOSTargetState)state
{
  return [FBCoreSimulatorNotifier resolveLeavesState:state forSimDevice:self.simulator.device];
}

#pragma mark Focus

- (FBFuture<NSNull *> *)focus
{
  // We cannot 'focus' a SimulatorApp for the non-default device set.
  NSString *deviceSetPath = self.simulator.customDeviceSetPath;
  if (deviceSetPath) {
    return [[FBSimulatorError
      describeFormat:@"Focusing on the Simulator App for a simulator in a custom device set (%@) is not supported", deviceSetPath]
      failFuture];
  }
  
  // Find the running instances of SimulatorApp.
  NSArray<NSRunningApplication *> *apps = NSWorkspace.sharedWorkspace.runningApplications;
  NSPredicate *simulatorAppPredicate = [NSPredicate predicateWithBlock:^(NSRunningApplication *application, NSDictionary<NSString *,id> *__) {
    return [application.bundleIdentifier isEqualToString:@"com.apple.iphonesimulator"];
  }];
  NSArray<NSRunningApplication *> *simulatorApps = [apps filteredArrayUsingPredicate:simulatorAppPredicate];

  // If we have no SimulatorApp running then we can instead launch one in a focused state
  if (simulatorApps.count == 0) {
    NSError *error = nil;
    NSRunningApplication *simulatorApp = [FBSimulatorLifecycleCommands launchSimulatorApplicationForDefaultDeviceSetWithError:&error];
    if (!simulatorApp) {
      return [FBFuture futureWithError:error];
    }
    return FBFuture.empty;
  }

  // Multiple apps, we don't know which to select.
  if (simulatorApps.count > 1) {
    return [[FBSimulatorError
      describeFormat:@"More than one SimulatorApp %@ running, focus is ambiguous", [FBCollectionInformation oneLineDescriptionFromArray:simulatorApps]]
      failFuture];
  }
  
  // Otherwise we have a single Simulator App to activate.
  NSRunningApplication *simulatorApp = simulatorApps.firstObject;
  if (![simulatorApp activateWithOptions:NSApplicationActivateIgnoringOtherApps]) {
    return [[FBSimulatorError
      describeFormat:@"Failed to focus %@", simulatorApp]
      failFuture];
  }

  return FBFuture.empty;
}

+ (NSRunningApplication *)launchSimulatorApplicationForDefaultDeviceSetWithError:(NSError **)error
{
  // Obtain the location of the SimulatorApp
  FBBundleDescriptor *applicationBundle = FBXcodeConfiguration.simulatorApp;
  NSURL *applicationURL = [NSURL fileURLWithPath:applicationBundle.path];

  // We only want to ever connect to the default SimulatorApp, including re-activating it rather than creating a new instance.
  NSError *innerError = nil;
  NSRunningApplication *application = [NSWorkspace.sharedWorkspace
    launchApplicationAtURL:applicationURL
    options:NSWorkspaceLaunchDefault
    configuration:@{}
    error:&innerError];

  if (!application) {
    return [[[FBSimulatorError
      describe:@"Failed to launch SimulatorApp"]
      causedBy:innerError]
      fail:error];
  }

  return application;
}

#pragma mark Connection

- (FBFuture<NSNull *> *)disconnectWithTimeout:(NSTimeInterval)timeout logger:(nullable id<FBControlCoreLogger>)logger
{
  NSDate *date = NSDate.date;
  return [[[self
    terminateConnections]
    timeout:timeout waitingFor:@"Simulator connections to teardown"]
    onQueue:self.simulator.workQueue map:^(id _) {
      [logger.debug logFormat:@"Simulator connections torn down in %f seconds", [NSDate.date timeIntervalSinceDate:date]];
      return NSNull.null;
    }];
}

- (FBFuture<NSNull *> *)terminateConnections
{
  FBSimulatorHID *hid = self.hid;
  FBSimulatorBridge *bridge = self.bridge;
  return [[FBFuture
    futureWithFutures:@[
      (hid ? [hid disconnect] : FBFuture.empty),
      (bridge ? [bridge disconnect] : FBFuture.empty),
    ]]
    onQueue:self.simulator.workQueue chain:^(FBFuture *_) {
      // Nullify
      self.hid = nil;
      self.bridge = nil;
      return FBFuture.empty;
    }];
}

#pragma mark Bridge

- (FBFuture<FBSimulatorBridge *> *)connectToBridge
{
  if (self.bridge) {
    return [FBFuture futureWithResult:self.bridge];
  }

  return [[FBSimulatorBridge
    bridgeForSimulator:self.simulator]
    onQueue:self.simulator.workQueue map:^(FBSimulatorBridge *bridge) {
      self.bridge = bridge;
      return bridge;
    }];
}

#pragma mark Framebuffer

- (FBFuture<FBFramebuffer *> *)connectToFramebuffer
{
  FBSimulator *simulator = self.simulator;
  return [FBFuture
    onQueue:simulator.workQueue resolveValue:^(NSError **error) {
      return [FBFramebuffer mainScreenSurfaceForSimulator:simulator logger:simulator.logger error:error];
    }];
}

#pragma mark Bridge

- (FBFuture<FBSimulatorHID *> *)connectToHID
{
  if (self.hid) {
    return [FBFuture futureWithResult:self.hid];
  }
  return [[FBSimulatorHID
    hidForSimulator:self.simulator]
    onQueue:self.simulator.workQueue map:^(FBSimulatorHID *hid) {
      self.hid = hid;
      return hid;
    }];
}

#pragma mark URLs

- (FBFuture<NSNull *> *)openURL:(NSURL *)url
{
  NSParameterAssert(url);
  NSError *error = nil;

  int retry = 0;
  do {
    // Retry openURL 2 times to alleviate Rosetta startup slowness.
    if ([self.simulator.device openURL:url error:&error]) {
      return [FBFuture futureWithResult:[NSNull null]];
    }
    retry++;
  } while (retry <= OPEN_URL_RETRIES);

  return [[[FBSimulatorError
    describeFormat:@"Failed to open URL %@ on simulator %@", url, self.simulator]
    causedBy:error]
    failFuture];
}

@end

File: FBSimulatorControl/Commands/FBSimulatorLifecycleCommands.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import AppKit
@preconcurrency import CoreSimulator
@preconcurrency import FBControlCore
@preconcurrency import Foundation

private let openURLRetries = 2

/// The ways simulator lifecycle operations can fail, as data rather than assembled strings.
public enum FBSimulatorLifecycleError: Error {
  case focusUnsupportedForCustomDeviceSet(deviceSetPath: String)
  case focusAmbiguous(runningApplications: String)
  case focusFailed(applicationDescription: String)
  case openURLFailed(url: URL, simulatorDescription: String, underlying: Error?)
}

extension FBSimulatorLifecycleError: LocalizedError {
  public var errorDescription: String? {
    switch self {
    case let .focusUnsupportedForCustomDeviceSet(deviceSetPath):
      return "Focusing on the Simulator App for a simulator in a custom device set (\(deviceSetPath)) is not supported"
    case let .focusAmbiguous(runningApplications):
      return "More than one SimulatorApp \(runningApplications) running, focus is ambiguous"
    case let .focusFailed(applicationDescription):
      return "Failed to focus \(applicationDescription)"
    case let .openURLFailed(url, simulatorDescription, _):
      return "Failed to open URL \(url) on simulator \(simulatorDescription)"
    }
  }
}

public final class FBSimulatorLifecycleCommands {

  // MARK: - Properties

  private weak var simulator: FBSimulator?
  private var hid: FBSimulatorHID?

  // MARK: - Initializers

  public class func commands(with simulator: FBSimulator) -> FBSimulatorLifecycleCommands {
    FBSimulatorLifecycleCommands(simulator: simulator)
  }

  private init(simulator: FBSimulator) {
    self.simulator = simulator
  }

  // MARK: - Async

  fileprivate func boot(_ configuration: FBSimulatorBootConfiguration) async throws {
    guard let simulator = self.simulator else {
      throw FBWeakTargetError.simulator
    }
    try await FBSimulatorBootStrategy.boot(simulator, with: configuration)
  }

  fileprivate func shutdown() async throws {
    guard let simulator = self.simulator else {
      throw FBWeakTargetError.simulator
    }
    try await FBSimulatorShutdownStrategy.shutdown(simulator)
  }

  fileprivate func reboot() async throws {
    try await shutdown()
    try await boot(FBSimulatorBootConfiguration.default)
  }

  fileprivate func erase() async throws {
    guard let simulator = self.simulator else {
      throw FBWeakTargetError.simulator
    }
    try await FBSimulatorEraseStrategy.erase(simulator)
  }

  fileprivate func resolveStateAsync(_ state: FBiOSTargetState) async throws {
    guard let simulator = self.simulator else {
      throw FBWeakTargetError.simulator
    }
    try await FBiOSTargetResolveState(simulator, state)
  }

  fileprivate func resolveLeavesStateAsync(_ state: FBiOSTargetState) async throws {
    guard let simulator = self.simulator else {
      throw FBWeakTargetError.simulator
    }
    try await bridgeFBFutureVoid(FBCoreSimulatorNotifier.resolveLeavesState(state, for: simulator.device))
  }

  fileprivate func focus() async throws {
    guard let simulator = self.simulator else {
      throw FBWeakTargetError.simulator
    }
    // The Simulator host app (Simulator.app, or DeviceHub.app on Xcode 27+) only displays
    // simulators in the default device set, so 'focus' is unsupported for a custom device set.
    // This is also why Xcode parallel testing — which clones into a non-default device set — is
    // not visible in DeviceHub (Apple known issue 176809181).
    if let deviceSetPath = simulator.customDeviceSetPath {
      throw FBSimulatorLifecycleError.focusUnsupportedForCustomDeviceSet(deviceSetPath: deviceSetPath)
    }

    // Find the running instances of the Simulator host app. Xcode 27 renamed Simulator.app
    // (com.apple.iphonesimulator) to DeviceHub.app (com.apple.dt.Devices); match either.
    let apps = NSWorkspace.shared.runningApplications
    let simulatorAppBundleIDs: Set<String> = ["com.apple.iphonesimulator", "com.apple.dt.Devices"]
    let simulatorApps = apps.filter { app in
      guard let bundleIdentifier = app.bundleIdentifier else { return false }
      return simulatorAppBundleIDs.contains(bundleIdentifier)
    }

    // If we have no SimulatorApp running then we can instead launch one in a focused state
    guard let simulatorApp = simulatorApps.first else {
      try await FBSimulatorLifecycleCommands.launchSimulatorApplicationForDefaultDeviceSet()
      return
    }

    // Multiple apps, we don't know which to select.
    if simulatorApps.count > 1 {
      throw FBSimulatorLifecycleError.focusAmbiguous(runningApplications: FBCollectionInformation.oneLineDescription(from: simulatorApps))
    }

    // Otherwise we have a single Simulator App to activate.
    if !simulatorApp.activate() {
      throw FBSimulatorLifecycleError.focusFailed(applicationDescription: String(describing: simulatorApp))
    }
  }

  private class func launchSimulatorApplicationForDefaultDeviceSet() async throws {
    let applicationBundle = FBXcodeConfiguration.simulatorApp
    let applicationURL = URL(fileURLWithPath: applicationBundle.path)
    let configuration = NSWorkspace.OpenConfiguration()
    configuration.activates = true
    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
      NSWorkspace.shared.openApplication(at: applicationURL, configuration: configuration) { _, error in
        if let error {
          continuation.resume(throwing: error)
        } else {
          continuation.resume()
        }
      }
    }
  }

  fileprivate func disconnect(withTimeout timeout: TimeInterval, logger: (any FBControlCoreLogger)?) async throws {
    guard self.simulator != nil else {
      throw FBWeakTargetError.simulator
    }
    let date = Date()
    let teardownFuture =
      fbFutureFromAsync { [self] in
        try await terminateConnections()
        return NSNull()
      }
      .timeout(timeout, waitingFor: "Simulator connections to teardown")
      .retyped(FBFuture<NSNull>.self)
    try await bridgeFBFutureVoid(teardownFuture)
    logger?.debug().log("Simulator connections torn down in \(Date().timeIntervalSince(date)) seconds")
  }

  private func terminateConnections() async throws {
    hid?.disconnect()
    self.hid = nil
  }

  fileprivate func connectToFramebuffer() async throws -> FBFramebuffer {
    guard let simulator = self.simulator else {
      throw FBWeakTargetError.simulator
    }
    return try FBFramebuffer.mainScreenSurface(for: simulator, logger: simulator.logger)
  }

  fileprivate func connectToHID() async throws -> FBSimulatorHID {
    if let hid = self.hid {
      return hid
    }
    guard let simulator = self.simulator else {
      throw FBWeakTargetError.simulator
    }
    let hid = try FBSimulatorHID(for: simulator)
    self.hid = hid
    return hid
  }

  fileprivate func open(_ url: URL) async throws {
    guard let simulator = self.simulator else {
      throw FBWeakTargetError.simulator
    }
    var lastError: NSError?
    for _ in 0...openURLRetries {
      lastError = nil
      do {
        try simulator.device.open(url)
        return
      } catch {
        lastError = error as NSError
      }
    }
    throw FBSimulatorLifecycleError.openURLFailed(url: url, simulatorDescription: String(describing: simulator), underlying: lastError)
  }
}

// MARK: - FBSimulator+LifecycleCommands

extension FBSimulator: LifecycleCommands {

  public func resolveState(_ state: FBiOSTargetState) async throws {
    try await lifecycle.resolveStateAsync(state)
  }

  public func resolveLeavesState(_ state: FBiOSTargetState) async throws {
    try await lifecycle.resolveLeavesStateAsync(state)
  }
}

// MARK: - FBSimulator+PowerCommands

extension FBSimulator: PowerCommands {

  public func shutdown() async throws {
    try await lifecycle.shutdown()
  }

  public func reboot() async throws {
    try await lifecycle.reboot()
  }
}

// MARK: - FBSimulator+EraseCommands

extension FBSimulator: EraseCommands {

  public func erase() async throws {
    try await lifecycle.erase()
  }
}

// MARK: - FBSimulator+SimulatorLifecycleCommands

extension FBSimulator: SimulatorLifecycleCommands {

  public func boot(_ configuration: FBSimulatorBootConfiguration) async throws {
    try await lifecycle.boot(configuration)
  }

  public func focus() async throws {
    try await lifecycle.focus()
  }

  public func disconnect(withTimeout timeout: TimeInterval, logger: (any FBControlCoreLogger)?) async throws {
    try await lifecycle.disconnect(withTimeout: timeout, logger: logger)
  }

  public func connectToFramebuffer() async throws -> FBFramebuffer {
    try await lifecycle.connectToFramebuffer()
  }

  public func open(_ url: URL) async throws {
    try await lifecycle.open(url)
  }

  public func connectToHID() async throws -> FBSimulatorHID {
    try await lifecycle.connectToHID()
  }
}
